### PR TITLE
FEATURE: RAIL-351 Attribute Elements component

### PR DESCRIPTION
**Pull request by tomasreichmann**

### CODE REVIEW

**Comment by tomasreichmann** [comment]
extended test - gemini

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/216/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-el7/848/ : FAILURE in 2m 00s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/163/ : FAILURE in 2m 11s (non-voting)

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/217/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/41/ : FAILURE in 3m 03s

**Comment by tomasreichmann** [comment]
extended test - cafe

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/218/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/42/ : FAILURE in 12s

**Comment by ocetnik** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/219/ : SUCCESS in 16s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/43/ : FAILURE in 3m 20s

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/220/ : SUCCESS in 15s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-el7/849/ : FAILURE in 39s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/164/ : FAILURE in 1m 05s (non-voting)

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/221/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/44/ : FAILURE in 2m 13s

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/11/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/223/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/45/ : SUCCESS in 3m 50s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/222/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-el7/850/ : FAILURE in 1m 35s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/165/ : FAILURE in 2m 06s (non-voting)

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/224/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-el7/851/ : FAILURE in 1m 54s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/166/ : FAILURE in 2m 00s (non-voting)

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/12/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/225/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/46/ : SUCCESS in 4m 02s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/226/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-el7/852/ : SUCCESS in 2m 06s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/167/ : SUCCESS in 2m 42s (non-voting)

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/13/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/227/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/47/ : SUCCESS in 4m 01s

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/228/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-el7/853/ : SUCCESS in 1m 55s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/168/ : SUCCESS in 3m 14s (non-voting)

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/14/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/229/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/48/ : SUCCESS in 3m 55s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/246/ : SUCCESS in 18s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/185/ : SUCCESS in 2m 42s

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/18/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/247/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/52/ : UNSTABLE in 3m 06s

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/249/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/188/ : SUCCESS in 3m 02s

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/19/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/250/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/53/ : UNSTABLE in 4m 06s

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/251/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/189/ : SUCCESS in 3m 15s

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/20/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/252/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/54/ : UNSTABLE in 2m 55s

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/253/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/190/ : SUCCESS in 3m 07s

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/21/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/254/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/55/ : UNSTABLE in 3m 03s

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/255/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/191/ : SUCCESS in 1m 34s

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/22/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/256/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/56/ : UNSTABLE in 3m 03s

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/257/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/192/ : SUCCESS in 2m 59s

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/23/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/258/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/57/ : SUCCESS in 2m 49s

File: .nvmrc [hunk 1]
```diff
@@ -0,0 +1 @@
+6.10.1
```
Comment (lhorak) [inline .nvmrc]: .nvmrc gitignore

File: src/components/filters/AttributeFilter/AttributeElements.tsx [hunk 2]
```diff
@@ -0,0 +1,231 @@
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+import * as GoodData from 'gooddata';
+import { get } from 'lodash';
+
+export interface IPaging {
```
Comment (lhorak) [inline src/components/filters/AttributeFilter/AttributeElements.tsx]: You could make this all required and use it on line 26-30

File: src/components/filters/AttributeFilter/AttributeElements.tsx [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        uri: null,
+        identifier: null,
+        options: null,
+        children: ({
+            validElements,
+            loadMore,
+            isLoading,
+            error
+        }) => {
+            const paging: IPaging = validElements ? validElements.paging : {};
```
Comment (lhorak) [inline src/components/filters/AttributeFilter/AttributeElements.tsx]: Change IPaging to Partial<IPaging> to make it all optional

File: .nvmrc (hunk 1 above)
Comment (lhorak) [reply .nvmrc]: or newline

**Comment by tomasreichmann** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/261/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/195/ : SUCCESS in 1m 33s

**Comment by osklive-kacatko** [comment]
Link to Storybook test result:

https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/24/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-opensource-repository-check/262/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-react-components-pull-request-dispatcher/58/ : SUCCESS in 3m 58s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/gooddata-react-components-merge-compliance-check/94/ : SUCCESS in 19s
- https://ci.intgdc.com/job/gooddata-react-components-unit-tests-zuul-docker-el7/198/ : SUCCESS in 2m 38s
- https://ci.intgdc.com/job/gooddata-react-components-storybook-zuul-docker/26/ : SUCCESS in 3m 27s

File: src/components/filters/AttributeFilter/AttributeElements.tsx [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+                    <pre>
+                        isLoading: {isLoading.toString()}
+                        offset: {offset}
+                        count: {count}
+                        total: {total}
+                        nextOffset: {nextOffset}
+                        validElements:
+                        {JSON.stringify(validElements, null, '\t')}
+                    </pre>
+                </div>
```
Comment (zbycz) [inline src/components/filters/AttributeFilter/AttributeElements.tsx]: Definitely remove "example children" from component`s implementation. Story is the right place - and there it is as well.

File: stories/AttributeFilter.tsx [hunk 5]
```diff
@@ -31,3 +36,72 @@ storiesOf('AttributeFilter', module)
...
+                    count: {count}<br />
+                    total: {total}<br />
+                    nextOffset: {offset + count}
+                </p>
+                <div>
+                    {validElements ? validElements.items.map((item: any) => (
+                        <span key={item.element.uri} style={{float: 'left'}}><AttributeFilterItem
+                            item={{
+                                selected: false,
+                                onSelect: action('select'),
```
Comment (zbycz) [inline stories/AttributeFilter.tsx]: Big like 👍

File: stories/AttributeFilter.tsx [hunk 6]
```diff
@@ -31,3 +36,72 @@ storiesOf('AttributeFilter', module)
...
+        validElements,
+        loadMore,
+        isLoading,
+        error
+    }: any) => {
+        const {
+            offset = null as any,
+            count = null as any,
+            total = null as any
+        } = validElements ? validElements.paging : {};
```
Comment (zbycz) [inline stories/AttributeFilter.tsx]: If you added `paging` parameter to the children function, this footprint would be much easier.

File: src/components/filters/AttributeFilter/AttributeElements.tsx [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    public render() {
+        const { validElements, isLoading, error } = this.state;
+        return this.props.children({
+            validElements,
+            loadMore: this.loadMore,
+            isLoading,
+            error
+        });
```
Comment (zbycz) [inline src/components/filters/AttributeFilter/AttributeElements.tsx]: Why object? I would prefer simple parameters. This adds unnecessary complexity and also i cant see any usecase (ie skiping one of the params in consumer / too many of them).

**Pull request merged**